### PR TITLE
refactor(compiler): initial changes to compute defer block dependencies

### PR DESCRIPTION
**Pull request by AndrewKushnir**
⚠️ **Disclaimer** ⚠️ **this PR implements syntax that is still in an open RFC. It will be adjusted once the RFC is closed.**

This PR contains several commits (see each commit for additional info) that add initial version of the logic to compute dependencies of defer blocks used in a template. The code contains a few TODOs that will be resolved in followup PRs.

## Does this PR introduce a breaking change?

- [ ] Yes
- [x] No

### CODE REVIEW

File: packages/compiler/src/render3/view/template.ts [hunk 1]
```diff
@@ -1071,8 +1073,53 @@ export class TemplateDefinitionBuilder implements t.Visitor<void>, LocalResolver
...
+      deferredDeps.forEach(deferredDep => {
+        if (deferredDep.isDeferrable) {
+          // Callback function, e.g. `function(m) { return m.MyCmp; }`.
+          const innerFn = o.fn(
+              [new o.FnParam('m', o.DYNAMIC_TYPE)],
+              [new o.ReturnStatement(o.variable('m').prop(deferredDep.symbolName))]);
+
+          const fileName = deferredDep.importPath!;
+          // Dynamic import, e.g. `import('./a').then(...)`.
+          const importExpr = (new o.DynamicImportExpr(fileName)).prop('then').callFn([innerFn]);
```
Comment (alxhub) [inline packages/compiler/src/render3/view/template.ts:1100]: Shouldn't we coalesce multiple imports from the same location?

File: packages/compiler/src/render3/view/t2_binder.ts [hunk 2]
```diff
@@ -266,7 +269,7 @@ class DirectiveBinder<DirectiveT extends DirectiveMeta> implements Visitor {
     // Next, use the `SelectorMatcher` to get the list of directives on the node.
     const directives: DirectiveT[] = [];
     this.matcher.match(cssSelector, (_selector, results) => directives.push(...results));
-    if (directives.length > 0) {
+    if (directives.length > 0 && !this.isInDeferBlock) {
```
Comment (alxhub) [inline packages/compiler/src/render3/view/t2_binder.ts:272]: This seems odd to me - the binder shouldn't be ignoring directives inside defer blocks completely. That's going to mess up future processing and responses, like the below logic which tries to determine the target of local references.

**Review by alxhub — CHANGES_REQUESTED** [review]

File: packages/compiler/src/render3/view/template.ts (hunk 1 above)
Comment (AndrewKushnir) [reply packages/compiler/src/render3/view/template.ts:1100]: Thanks Alex! Yes, we should coalesce multiple imports. I'll take a note on it and will create a followup PR with the improvement (to keep things simple in TDB in this PR).

File: packages/compiler/src/render3/view/t2_binder.ts (hunk 2 above)
Comment (AndrewKushnir) [reply packages/compiler/src/render3/view/t2_binder.ts:272]: This check prevents adding found directives to the set that is later returned from `getUsedDirectives()` function (and later treated as "eager" in the `.../annotations/component/src/handler.ts`). The logic below in this function still has access to all matched directives (from the local `directives` set).

Comment (AndrewKushnir) [reply packages/compiler/src/render3/view/t2_binder.ts]: As we discussed offline, I've pushed one more commit with the changes to retain an old interface of the DirectiveBinder class and introduce 2 new methods to retrieve eagerly used directives and pipes: https://github.com/angular/angular/pull/51162/commits/7faa2264aac11fa4b533130646c58f7de3085b40.

**Comment by AndrewKushnir** [comment]
[Exploratory presubmit](https://test.corp.google.com/OCL:551028495:BASE:551064988:1690338620872:afc1ea1e).

**Review by alxhub — APPROVED** [review]

**Comment by AndrewKushnir** [comment]
[Global presubmit](https://test.corp.google.com/OCL:551028495:BASE:552142744:1690663752138:20a2b7c3).

**Comment by AndrewKushnir** [comment]
Caretaker note: [the TGP is "green"](https://fusion2.corp.google.com/presubmit/551028495/OCL:551028495:BASE:552153337:1690670321007:452b13b9/targets), this PR is ready for merge.

**Comment by alxhub** [comment]
This PR was merged into the repository by commit efb486e8bc8712a961a73180fed324bbaac93348.

**Pull request closed without merging**